### ISSUE TITLE
Pressure Level Embeddings

### ISSUE THREAD

**Issue by mc4117** [ecmwf/anemoi-core#25]
### Is your feature request related to a problem? Please describe.

The current code does not take into account embeddings in the z-dimension (either through height or pressure levels). We saw with Sara's work that pressure level weightings can have a large impact on results, especially in the upper atmosphere. 

This has been done in Aurora (https://arxiv.org/abs/2405.13063) which shows significant improvements to graphcast in the upper atmosphere, and also in https://arxiv.org/abs/2405.06004v1 and https://arxiv.org/pdf/2405.14527. 

### Describe the solution you'd like

Include pressure level embeddings in our encoder implementation and evaluate the impact of this.

### Organisation

ECMWF

**Comment by mc4117** [comment]
Ana and I have been making some progress on this ticket. We first tried a naive approach of adding the pressure levels or concatenating the pressure levels to the existing tensors
https://confluence.ecmwf.int/display/DWF/Pressure+Level+Embeddings

This did not degrade the model but it also did not improve the model for the tests we carried out.

We are increasingly thinking that we will have to add a vertical dimension to the model if we want to do this properly.

For now, we are thinking of a simpler version of the model which just includes pressure levels (we can then think about how to add the surface back in later).

Our current idea would be (before the encoder) to reshape the data into (-, 78) -> (-, 13, 6)
1. We could try and pass this object directly to the encoder but we think that adding this dimension will break the mapper.
2. We could pass the (-, 13, 6) through a linear layer to compress it to a (-, 1, 6) object which could then be concatenated to the surface variables
3. We could pass the (-, 13, 6) object through a Perceiver (basically copying aurora). This last point could be very linked to https://github.com/ecmwf/aifs-dev/issues/15 (@jakob-schloer)

Thought and contributions welcome!

(I think there may have been a mistake and the two commits Helen linked are not related to this issue. We have a branch of anemoi-core but we have not developed much on this https://github.com/ecmwf/anemoi-core/tree/feature/embeddings.)

**Comment by ssmmnn11** [comment]
Having a quick look at aurora, I think what they do:

aurora allows for a different number of vertical levels ; to achieve that the model has to know what level the data is on. We currently have the vertical dimension explicitly because we have column vectors. 

it seems to me that to become flexible, aurora creates 2d patches of each variable at one pressure level, e.g. 500hPa, and embeds them before summing them together. Then, to "know" what level it is, it creates a vertical coordinate embedding (fourier emb) which is added to the summed patch embedding, same as one would do with add. position embeddings. this is (probably) important for the 3d perciever, but in our case I don't know if this will make an impact because of the fixed number our input / output levels we have.

**Comment by mc4117** [comment]
By the column vector do you mean the objects with shape (batch ensemble grid) (time vars)?

**Comment by ssmmnn11** [comment]
yes, we have a fixed number of input and output variables which are stacked on top of each other. and these do not change. hence the model can assume that the information at a certain position always relates to the same level. If you would e.g. randomly change the ordering (could be a nice test - create random indices for input and output fields), the I would assume the model would struggle and would need additional information, i.e. tagging with pressure levels.
